src/pybindings/xcdf/xcdf.cpp

```diff
@@ -0,0 +1,227 @@
+
+    field7 << 0.12;
+
+    field8 << 2 << 2 << 1 << 1;
+    field9 << 1. << 2. << 3. << 4. << 5. << 6.;
+
+    std::cout << f.Write() << std::endl;
+
+    for (int k = 0; k < 1000; k++)
+    {
+        field1 << 2;
+        field2 << 1 << 3;
+        field3 << 0.3;
+        field4 << std::numeric_limits<double>::signaling_NaN();
+        field5 << std::numeric_limits<double>::infinity();
+        field6 << 0xDEADBEEFDEADBEEFULL;
+        field7 << 0.12;
+        field8 << 2 << 2 << 1 << 1;
+        field9 << 1. << 2. << 3. << 4. << 5. << 6.;
+        f.Write();
+    }
+
+    f.AddComment("test file");
+
+    // Trailer alias
+    f.CreateAlias("testTrailerAlias", "double(testAlias + 2)");
+
+    f.Close();
+
+    return 0;
+}
+
+struct DictBuilder
+{
+    py::dict data;
+
+    template <typename T>
+    void operator()(const XCDFField<T> &field)
+    {
+
+        py::str key = field.GetName();
+        // return vector fields as numpy arrays
+        if (field.HasParent())
+        {
+            py::array_t<T> array(field.GetSize());
+            py::buffer_info buffer = array.request();
+            T *ptr = static_cast<T *>(buffer.ptr);
+
+            for (size_t i = 0; i < field.GetSize(); i++)
+            {
+                ptr[i] = field[i];
+            }
+
+            data[key] = array;
+            // scalar fields as python objects (None if not set)
+        }
+        else
+        {
+            if (field.GetSize() == 0)
+            {
+                data[key] = py::none();
+            }
+            else
+            {
+                data[key] = field[0];
+            }
+        }
+    }
+};
+
+PYBIND11_MODULE(xcdf, m)
+{
+    m.doc() = "Python Bindings for XCDF";
+    m.attr("__version__") = getVersion();
+
+    m.def("write_test_file", &write_test_file, "Write an XCDF test file in C++");
+
+    py::class_<XCDFFieldDescriptor>(m, "FieldDescriptor")
+        .def("__repr__", [](XCDFFieldDescriptor &self)
+             {
+            std::stringstream ss;
+            ss << "FieldDescriptor(name=" << self.name_
+                << ", type=" << static_cast<int>(self.type_)
+                << ", parent=" << self.parentName_
+                << ", raw_resolution=" << self.rawResolution_
+                << ")";
+
+            return ss.str(); })
+        .def_property_readonly("name", [](XCDFFieldDescriptor &self)
+                               { return self.name_; })
+        .def_property_readonly("type", [](XCDFFieldDescriptor &self)
+                               { return self.type_; })
+        .def_property_readonly("parent", [](XCDFFieldDescriptor &self)
+                               { return self.parentName_; })
+        .def_property_readonly("raw_resolution", [](XCDFFieldDescriptor &self)
+                               { return self.rawResolution_; });
+
+    py::class_<XCDFFile>(m, "File")
+        .def(py::init<const char *, const char *>())
+        .def("close", &XCDFFile::Close)
```

Comment:
to make things nicer, you can provide docstrings and `py::arg` specifications for all the calls here.

Reply:
I didn't add docstrings because the pybind11 talks only about sphinx and I was actually planning to use doxygen (I never used sphinx outside of python)

regarding `py::arg` you mean [this](https://pybind11.readthedocs.io/en/stable/basics.html?highlight=py%3A%3Aarg#keyword-arguments)?

Reply:
Docstrings are needed! They are shown within python, e.g. when doing `help(<thing>)` or when looking at reprs, right now it just shows:

```
In [1]: from xcdf import File

In [2]: File
Out[2]: xcdf.xcdf.File

In [3]: File?
Docstring:      <no docstring>
Init docstring: __init__(self: xcdf.xcdf.File, path: str, mode: str = 'r') -> None
File:           ~/Projects/XCDF/src/pybindings/xcdf/xcdf.cpython-310-x86_64-linux-gnu.so
Type:           pybind11_type
Subclasses:    
```

Reply:
> Docstrings are needed! 

of course :)

How do you write docstrings in an sowftware independend way in this .cpp file?
can I still do it as they show [here](https://pybind11.readthedocs.io/en/stable/advanced/misc.html?highlight=docstring#generating-documentation-using-sphinx) even though I am not sure if sphinx will be used in the end?

Reply:
You need to write them for python! So use the usual syntax. That is independent of whether they in the end are rendered using sphinx.

They are just given as argument to the pybind11 methods.

Reply:
done